utils/github: fix broken pipe error

UniqMartin opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/.github/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Steps to reproduce:

- Unset any possibly set `HOMEBREW_GITHUB_*` environment variables.
- Switch to a Git without `git-credential-osxkeychain` or temporarily rename this binary.

Before:

```console
$ brew search php
Error: Broken pipe
```

After:

```console
$ brew search php
No formula found for "php".
==> Searching pull requests...
Closed pull requests:
n98-magerun: migrate to php (https://github.com/Homebrew/homebrew-core/pull/1534)
Error: GitHub API Error: API rate limit exceeded for A.B.C.D. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.)
Try again in 25 minutes and 34 seconds, or create a personal access token:
  https://github.com/settings/tokens/new?scopes=&description=Homebrew
and then set the token as: export HOMEBREW_GITHUB_API_TOKEN="your_new_token"
```

Closes #573.

**Update:** Slightly expanded the scope of the PR [as mentioned in a comment below](https://github.com/Homebrew/brew/pull/581#issuecomment-234932442).

xu-cheng commented:
Can we do:
```ruby
begin
 github_credentials = Utils.popen ...
rescue
end
```

Or
```ruby
 github_credentials = Utils.popen (...) rescue ""
```

ilovezfs commented:
Would this be solved by the proposal to vendor git?

UniqMartin commented:
> Can we do:

Yes, if that seems preferable. The current version is optimized for a minimal diff (no need for a dedicated `begin` block and adjustment of the indentation) and since no other part of the code is doing I/O it's essentially the same. (But I can definitely see the benefit of bringing the `rescue` closer to the source of the exception.)

> Or

No. RuboCop objects to `rescue` as a modifier and I agree. It's too important to be stuffed away at the end of the line where it's easier to overlook and the modifier doesn't allow to specify the exceptions that should be handled, making it more likely to unintentionally handle other unrelated exceptions that shouldn't be silently swallowed.

> Would this be solved by the proposal to vendor git?

I think vendoring Git would reduce the likelihood of this happening, but I'm pretty sure there can be other error conditions that lead to an early exit without accepting any input even if the helper is present. (And I think the vendored Git was only meant to be used if the host Git is deemed too old, so it wouldn't actually help in many cases—unless I'm misremembering this part.)

ilovezfs commented:
It seems odd that we're getting a broken pipe instead of an explicit error? Is this caused by some fragility in Utils.popen?

UniqMartin commented:
Our version of `popen` redirects standard error to `/dev/null`, thus any error messages will be discarded, but the result of the invocation (the exit code) will still be reflected in `$?` after the block given to `popen` terminates.

The process we're talking to is free to terminate or stop accepting input at any time, thus the broken pipe error can literally happen in the middle of a `puts`/`write` operation and (to my knowledge) there isn't a more elegant/correct way of handling this than rescuing `Errno::EPIPE`.

ilovezfs commented:
Would this happen if `git credential-osxkeychain get` was executed in a shell instead?

UniqMartin commented:
This entirely depends on the program that is feeding input to the `git credential-osxkeychain get` process and how it decided to handle or not handle this case. The default signal handler for `SIGPIPE` is to simply kill the program as soon as a `write` fails due to a broken pipe. The other option is to have a custom handler for `SIGPIPE` or to mask this signal. The latter causes `write` to return `EPIPE` which is translated to the exception handled here.

So yes, in some sense this is also happening if executed in a shell, but we don't usually see the broken pipe error as it isn't considered a true error in most circumstances. E.g. in `cat file.txt | head -n 1` it is actually desirable that `head` dies after it has output the first line, so that the pipe is broken and `cat` can learn its pointless to continue reading the input file. (Whether `cat` internally relies on the default `SIGPIPE` behavior or checks for the `EPIPE` return value I don't know.)

Because the Homebrew process is doing a lot more than just feeding input to that helper (unlike in the `cat` example above), it makes sense to actually handle the broken pipe and then move on instead of being killed by the OS.

ilovezfs commented:
Why it preferable to return `[]` rather than give an explicit, custom error to the users? Won't the empty search output be confusing when they know what they're searching for does indeed exist?

UniqMartin commented:
> Won't the empty search output be confusing when they know what they're searching for does indeed exist?

It's not the search result that is empty, but the API credentials. Consequently, the code will continue to make unauthenticated GitHub API requests and either return the desired search results or report back with an error that the rate limit has been exceeded. That error also suggests to create and set a token, thus a possibly incomplete search result shouldn't be causing much confusion—certainly less than the broken pipe error. (See my top comment for an example output for the rate limit case.)

> Why it preferable to return `[]` rather than give an explicit, custom error to the users?

Because it is just one of several options and not having the credentials available this way isn't a fatal condition. We can still try and often succeed with an unauthenticated request. Maybe it's an unusual configuration on macOS, but it's still perfectly legal to have a Git installation without this credentials helper and making its absence stop the rest of the processing in its tracks didn't feel appropriate. If we wanted to make the user aware of this issue, I think the best thing we can do is output a warning. (I'm happy to make adjustments to the error reporting if my assessment of the situation isn't as accurate as I believe it to be.)

ilovezfs commented:
That makes sense. However, at least on macOS, I'm not sure why we'd want to support a git lacking this feature. We could simply require the use of an appropriate git, no?

ilovezfs commented:
In any case, I think we should consider requiring a git that has this feature, but it seems fine to :ship: this since you were saying you believe this can occur in some other scenario besides lacking the helper too.

UniqMartin commented:
Updated!

1. Refactored the fix to address feedback on code style/organization and moved some work to a helper method to avoid excessively deep nesting due to another `begin` block.

2. Did some minor refactoring to simplify the extraction code.

3. Fixed wrong element order when using `HOMEBREW_GITHUB_API_{USERNAME,PASSWORD}`. (Currently undocumented and probably not very widely used, if nobody noticed that the password and the user name were swapped.)

I'm happy to move the 2nd and/or 3rd commit to a separate PR, but I think this PR still makes sense with these additional changes on top if reviewed commit by commit.

ilovezfs commented:
Should this go in a separate method, too?

```
+    credentials = Hash[credentials.scan(/^(username|password)=(.*)$/)]
+    if credentials.key?("username") && credentials.key?("password")
+      credentials.values_at("password", "username")
+    else
+      []
+    end
```

UniqMartin commented:
How would further splitting this up benefit the code?

ilovezfs commented:
The rescue still seems to be a bit separated from where @xu-cheng was actually saying he'd prefer it be, no?

MikeMcQuaid commented on Library/Homebrew/utils/github.rb:
```diff
@@ -50,24 +50,34 @@ def api_credentials
     end
   end
 
+  def api_credentials_from_keychain(protocol, host)
```
I wonder if we need these args given that this only ever makes sense for HTTPS and GitHub.

UniqMartin replied:
They made it easier to locally test my fix because I don't have the GitHub credentials in my Keychain, but I do have other credentials accessible by the Git credentials helper. That said, I'll remove the arguments.

MikeMcQuaid replied:
Making it easier to locally test your fix is a good enough reason for me 👍

MikeMcQuaid commented on Library/Homebrew/utils/github.rb:
```diff
@@ -50,24 +50,34 @@ def api_credentials
+      pipe.read
+    end
+
+    credentials = Hash[credentials.scan(/^(username|password)=(.*)$/)]
```
I think the `Hash[]` approach is less obvious in what it's actually doing, personally.

UniqMartin replied:
I'll drop the refactoring commit and revert to the previous code.

MikeMcQuaid commented on Library/Homebrew/utils/github.rb:
```diff
@@ -50,24 +50,34 @@ def api_credentials
         ENV["HOMEBREW_GITHUB_API_TOKEN"]
       elsif ENV["HOMEBREW_GITHUB_API_USERNAME"] && ENV["HOMEBREW_GITHUB_API_PASSWORD"]
-        [ENV["HOMEBREW_GITHUB_API_USERNAME"], ENV["HOMEBREW_GITHUB_API_PASSWORD"]]
+        [ENV["HOMEBREW_GITHUB_API_PASSWORD"], ENV["HOMEBREW_GITHUB_API_USERNAME"]]
```
Why reverse the order, here?

UniqMartin replied:
Because this fixes a bug. Please see the commit message of the commit that changes this and the way the return value of `api_credentials` is used elsewhere.

MikeMcQuaid replied:
Cool. Would be good to put stuff like that in the PR body too, I think.

UniqMartin replied:
I did in https://github.com/Homebrew/brew/pull/581#issuecomment-234932442 and updated the PR body to reference this comment. What else could/should I have done? (Only noticed the bug after addressing the first feedback.)

MikeMcQuaid replied:
That's more than enough, I missed it, thanks!

MikeMcQuaid commented:
A few comments otherwise 👍 from me.

UniqMartin commented:
Updated to address all comments and to drop the refactoring due to questionable benefit.

MikeMcQuaid commented:
LGTM 👍

UniqMartin commented:
> The rescue still seems to be a bit separated from where @xu-cheng was actually saying he'd prefer it be, no?

@ilovezfs Thanks for reminding me of that! Are you happy with the current simplified diff?

ilovezfs commented:
@UniqMartin this looks great.

UniqMartin commented:
Thanks everyone for your feedback and your patience while I iterated through some revisions of this PR that were less than ideal! I'm now much more happy with the outcome.
